src/backend/distributed/commands/database.c

```diff
@@ -242,3 +284,187 @@ PreprocessAlterDatabaseSetStmt(Node *node, const char *queryString,
 
 	return NodeDDLTaskList(NON_COORDINATOR_NODES, commands);
 }
+
+
+/*
+ * PostprocessCreatedbStmt is executed after the statement is applied to the local
+ * postgres instance. In this stage we can prepare the commands that need to be run on
+ * all workers to create the database.
+ *
+ */
+List *
+PostprocessCreateDatabaseStmt(Node *node, const char *queryString)
+{
+	if (EnableCreateDatabasePropagation)
+	{
+		EnsureCoordinator();
+	}
+
+	if (!EnableCreateDatabasePropagation || !ShouldPropagate())
+	{
+		return NIL;
+	}
+
+	CreatedbStmt *stmt = castNode(CreatedbStmt, node);
+	char *databaseName = stmt->dbname;
+	bool missingOk = false;
+	Oid databaseOid = get_database_oid(databaseName, missingOk);
+
+	/*
+	 * TODO: try to reuse regular DDL infrastructure
+	 *
+	 * We do not do this right now because of the AssignDatabaseToShard at the end.
+	 */
+	List *workerNodes = TargetWorkerSetNodeList(NON_COORDINATOR_METADATA_NODES,
+												RowShareLock);
+	if (list_length(workerNodes) > 0)
+	{
+		char *createDatabaseCommand = DeparseTreeNode(node);
+
+		StringInfo internalCreateCommand = makeStringInfo();
+		appendStringInfo(internalCreateCommand,
+						 "SELECT pg_catalog.citus_internal_database_command(%s)",
+						 quote_literal_cstr(createDatabaseCommand));
+
+		/*
+		 * For the moment, we run CREATE DATABASE in 2PC, though that prevents
+		 * us from immediately doing a pg_dump | pg_restore when dealing with
+		 * a remote template database.
+		 */
+		bool outsideTransaction = false;
+
+		List *taskList = CreateDDLTaskList(internalCreateCommand->data, workerNodes,
+										   outsideTransaction);
+
+		bool localExecutionSupported = false;
+		ExecuteUtilityTaskList(taskList, localExecutionSupported);
+	}
+
+	/* synchronize pg_dist_object records */
+	ObjectAddress dbAddress = { 0 };
+	ObjectAddressSet(dbAddress, DatabaseRelationId, databaseOid);
+	MarkObjectDistributed(&dbAddress);
+
+
+	return NIL;
+}
+
+
+/*
+ * citus_internal_database_command is an internal UDF to
+ * create/drop a database in an idempotent maner without
+ * transaction block restrictions.
+ */
+Datum
+citus_internal_database_command(PG_FUNCTION_ARGS)
```

Comment:
@marcoslot, why we cannot do the same trick that we did for CREATE INDEX CONCURRENTLY etc., instead of using a UDF to call the internal funcs on workers :) ?